### ISSUE TITLE
`dask.array` module should expose dtype objects and  the `isdtype` function

### ISSUE THREAD

**Issue by ogrisel** [dask/dask#10387]
```python
>>> import dask.array as da
>>> da.float32
Traceback (most recent call last):
  Cell In[5], line 1
    da.float32
AttributeError: module 'dask.array' has no attribute 'float32'
>>> da.isdtype
Traceback (most recent call last):
  Cell In[6], line 1
    da.isdtype
AttributeError: module 'dask.array' has no attribute 'isdtype'
```

See:

- https://data-apis.org/array-api/latest/API_specification/generated/array_api.isdtype.html
- https://data-apis.org/array-api/latest/API_specification/data_types.html

It's not clear if the dtype objects should be all exposed as attributes of the top level module but this usually the case in other implementations of the spec:

```python
>>> import numpy.array_api as xp
<ipython-input-7-e68880e64b48>:1: UserWarning: The numpy.array_api submodule is still experimental. See NEP 47.
  import numpy.array_api as xp
>>> xp.float32
dtype('float32')
>>> xp.float64
dtype('float64')
>>> xp.isdtype(xp.zeros(shape=5, dtype=xp.float32), "real floating")
True
>>> numpy.array_api.float32 == numpy.float32
True
```

Since dask arrays use numpy arrays by default, I think that dask could just alias `numpy.float32` as `dask.array.float32` (and similarly for the other numpy dtypes).

Unfortunately, `numpy.isdtype` does not exist yet. Maybe `dask.array` could lazily alias `numpy.array_api.isdtype` as `dask.array.isdtype` with some module level `__getattr__` to not raise the `numpy.array_api` experimental warning by default.

Alternative, `dask.array` could ship its own implementation of `isdtype` (maybe vendoring/forking the one from `numpy.array_api`).

**Comment by lucascolley** [comment]
FWIW, exposing these would help with early adoption of Dask into SciPy. Now that we have `array_api_compat.dask`, we should be able to get our array-agnostic code working.

The slight problem at the minute is that in our tests, we make some assumptions about very basic things that NumPy, CuPy, PyTorch and the standard all have in common currently, such as `xp.float64`.

I understand that resolution to this may be somewhat blocked on a resolution to whether converging the main namespace or adding a separate namespace is desirable in gh-8750. For now though, even just exposing the dtypes as @ogrisel suggested would be helpful. Thanks!

**Comment by lucascolley** [comment]
bump on this!

**Comment by tylerjereddy** [comment]
@mrocklin Any suggestions on what might happen on the `dask` side?

**Comment by lucascolley** [comment]
The `np` dtypes have been exposed in the `dask.array` namespace in gh-11178. There isn't an `isdtype` yet - is that looking like a major problem for scikit-learn @ogrisel ? I'm going to run the next Dask nightly through the SciPy test suite and see if there is anything else that would be helpful.

**Comment by ogrisel** [comment]
In scikit-learn we do use the isdtype function but we now also depend on array-api-compat so this should not be a problem. The main integration problems are related to value dependent output shapes (nan in shapes) and the lack of a standardized way to trigger computation on array API namespaces with lazy evaluation as far as i remember but i have not checked again recently.

**Comment by lucascolley** [comment]
Do you test with namespaces wrapped by array-api-compat though? We test with the unwrapped namespaces in SciPy to ensure that our `array_namespace` is doing its job properly.

**Comment by ogrisel** [comment]
We only test with `array-api-compat` wrapping in our array API test harness. However we do not include dask yet. @lithomas1 started a PR to run tests with dask on our main repo and identify what was missing with the current state of array API support in scikit-learn:

- https://github.com/scikit-learn/scikit-learn/pull/28588

This effort was paused because we did not really see how to handle the lazy evaluation semantics in a library agnostic way. There is a summary and discussion of the main blockers here:

- https://github.com/scikit-learn/scikit-learn/issues/26724#issuecomment-2038674794

More anecdotally, there is also another missing feature in dask that makes some test fail (but in practice it wouldn't matter much to most of our users, I believe):

- https://github.com/scikit-learn/scikit-learn/pull/28588

**Comment by lucascolley** [comment]
> This effort was paused because we did not really see how to handle the lazy evaluation semantics in a library agnostic way.

Yeah, I just hit this: https://github.com/data-apis/array-api-compat/issues/175

> More anecdotally, there is also another missing feature in dask that makes some test fail (but in practice it wouldn't matter much to most of our users, I believe):

was this supposed to link to a specific comment perhaps, rather than the PR?

**Comment by lithomas1** [comment]
Thanks for the ping!

It's been a while since I worked on Dask Array API support (I'm hoping to have some more time for it by end of summer), but I'll try to write down what I remember here.

In addition to what was mentioned above, I think the biggest blocker is probably missing functionality in the dask.array linalg module as scikit-learn uses a lot of linalg functions. This prevents us from supporting things like PCA (which basically works with the array API compat stuff, but cannot be fully supported because we the ``score_samples`` method on the PCA estimator uses slogdet, which dask doesn't implement). The biggest missing feature linalg-wise I think for scikit-learn is probably eigenvalue/vector stuff (e.g. the eig family of methods).

r.e. laziness 
The places where data-dependent output shapes come into play (this is when you need to think about lazy stuff), are usually caused by a call to unique, and accessing like shape properties afterwards.
(I think scikit-learn does this to compute the number of classes)

Doing a setitem with an array with an unknown shape, might be work-aroundable with clever use of `xp.where` sometimes.
